obsolete/man/newaliases.1
.TH s/qmail: newaliases 1
.SH NAME
newaliases \- create a forwarding database from /etc/aliases
.SH SYNOPSIS
.B newaliases
.SH DESCRIPTION
.B newaliases
reads a table of
sendmail-style
forwarding instructions from
.B /etc/aliases
and converts them into a forwarding database in
.BR /etc/aliases.cdb .
The forwarding database can be used by
.BR fastforward .

For safety,
.B newaliases
writes the forwarding database to
.B /etc/aliases.tmp
and then moves
.B /etc/aliases.tmp
to
.BR /etc/aliases.cdb .
If there is a problem creating
.BR /etc/aliases.tmp ,
.B newaliases
complains and leaves
.B /etc/aliases.cdb
alone.
Deliveries can continue using
.B /etc/aliases.cdb
in the meantime.

.B newaliases
always creates
.B /etc/aliases.cdb
world-readable.

.B newaliases
makes no attempt to protect against
simultaneous updates of
.BR /etc/aliases.cdb .
.SH "INSTRUCTION FORMAT"
.B newaliases
imitates
sendmail's
handling of
.BR /etc/aliases .
For example,

.EX
   root: alice, bill
.EE

says that mail for
.B root
should be forwarded to
.B alice
and
.BR bill .

.B COMPATIBILITY WARNING:
.B newaliases
does not support file deliveries.
You can use the file delivery mechanism described in
.B dot-qmail(5)
instead.
.SH "SIMPLE ALIASES"
The simplest type of forwarding instruction
is a line of the form

.EX
   alias: recip
.EE

Any message sent to
.I alias
will be forwarded to the recipient address
.IR recip .
Addresses are compared to
.I alias
without regard to case.

Forwarding instructions are cumulative.
If
.I recip
is itself an alias,
messages to
.I alias
will be forwarded the same way as
messages to
.IR recip .
For example, with the following instructions,
messages to
.B postmaster@heaven.af.mil
or
.B root@heaven.af.mil
will be delivered to Bob:

.EX
   postmaster@heaven.af.mil: bob@heaven.af.mil
.EE
.br
.EX
   root@heaven.af.mil: postmaster@heaven.af.mil
.EE

.B COMPATIBILITY WARNING:
With
sendmail,
entries in
.B /etc/aliases
can override usernames.
With
.BR s/qmail ,
if you install
.B fastforward
in
.BR ~alias/.qmail-default ,
it will not see addresses that are controlled by other users.
See
.BR qmail-getpw (8).
To change this, see
.BR qmail-users (5).

.B COMPATIBILITY WARNING:
Various versions of
sendmail
do various strange things with circular alias definitions.
See
.BR setforward (1)
for details on
.BR fastforward 's
behavior.

.B COMPATIBILITY WARNING:
If there are several forwarding instructions for a single
.IR alias ,
sendmail
will complain;
.B fastforward
will silently use the first instruction.
.SH "WILDCARDS"
.I alias
can have the form
.I user@host.dom
for one user at one host,
.I @host.dom
for all users at one host, or
.I user
for one user at all hosts.

.B COMPATIBILITY WARNING:
sendmail
supports only
.IR user ;
it does not support per-host aliases.
It accepts
.I user@host.dom
if
.I host.dom
is a local host,
but it then treats it the same way as
.IR user ,
applying to all local hosts and virtual domains.
.SH "ADDRESS FORMATS"
Addresses in
.B /etc/aliases
are parsed the same way as addresses in RFC 822 message headers.
Parenthesized comments and bracketed addresses are permitted:

.EX
   root: bob (Bob, the postmaster)
   joe: Joe Shmoe <shmoe@heaven.af.mil>
.EE

Addresses with special characters must be quoted:

.EX
   fred: "spaced out mailbox"@heaven.af.mil
.EE

Address groups are not permitted,
since colons have a different use in
.BR /etc/aliases .

Any recipient address without a fully qualified domain name is
fed through the
.BR defaulthost ,
.BR defaultdomain ,
and
.B plusdomain
mechanisms described in
.BR qmail-header (5).

.B COMPATIBILITY WARNING:
sendmail's
handling of quotes and backslashes violates RFC 821 and RFC 822,
and is not supported by
.BR newaliases .
The
.B qmail-local
delivery mechanism
lets each user manage several addresses,
so there is no need for a special syntax to get around forwarding.
.SH "MULTIPLE RECIPIENTS"
An instruction may list more than one recipient address:

.EX
   alias: recip1, recip2, recip3
.EE

Any message sent to
.I alias
will be forwarded to all of the addresses.

A forwarding instruction may be split across several lines.
Each line past the first must either (1) begin with space or tab
or (2) be empty:

.EX
   hostmaster:
.EE
.br
.EX
      fred,
.EE
.br
.EX
      joe
.EE

.B COMPATIBILITY WARNING:
sendmail
requires the colon to be on the first line
of a multi-line forwarding instruction.
.B newaliases
doesn't care whether the colon is present at all.

.B COMPATIBILITY WARNING:
sendmail
does not permit blank lines in the middle of continuations.
This has the undesirable effect that a blank line behaves differently
from a line containing a single space.
.SH "COMMENTS"
Any line in
.B /etc/aliases
that begins with # is ignored:

.EX
   # this is a comment
.EE

A comment may be split across several lines.
Each line past the first must either (1) begin with space or tab
or (2) be empty.

.B COMPATIBILITY WARNING:
sendmail
does not permit continuations of comment lines.
.SH "PROGRAMS"
If a recipient address does not contain a domain name,
and begins with a vertical bar,
.B newaliases
takes the rest of the address as a program to run:

.EX
   weather: "|weather-server"
.EE

.B fastforward
will run
.B weather-server
when a message arrives for
.BR weather .

.B COMPATIBILITY WARNING:
Internet addresses can legitimately start with
a slash or vertical bar.
.B newaliases
treats anything with an unquoted @ as an address.
sendmail appears to have various problems
coping with these addresses,
and with commands that contain @ signs.

.B COMPATIBILITY WARNING:
.B newaliases
does not allow a vertical bar before double quotes.
.SH "INCLUDE FILES"
A recipient address of the form
.B :include:\fIfile
means ``every address listed in
.IR file .''
(Actually
.B fastforward
reads
.IR file\fB.bin ;
see
.BR newinclude (1)
for further details.)

Note that
.I file
is read by
.BR fastforward ,
not
.BR newaliases ,
so the system administrator does not have to run
.B newaliases
every time
.I file
changes.
.I file
must be world-readable
and accessible to
.BR fastforward .

.B COMPATIBILITY WARNING:
If an
.B :include:
file is unreadable or nonexistent,
sendmail
skips it;
.B fastforward
defers delivery of the message.

.B COMPATIBILITY WARNING:
sendmail
does not permit spaces inside the literal text
.BR :include: .
.B newaliases
does.

.B COMPATIBILITY WARNING:
Versions of
sendmail
before V8 did not strip quotes from
.B :include:
filenames.
.SH "ALIAS OWNERS"
If there is an alias for
.BR owner-\fIlist ,
any message forwarded through
.I list
will have its envelope sender set to
.BR owner-\fIlist ,
so that bounces go back to
.BR owner-\fIlist .

.B COMPATIBILITY WARNING:
When an alias includes the same recipient both inside and outside
a mailing list,
.B fastforward
sends the message twice,
once with each envelope sender.
sendmail
sends the message only once;
its choice of envelope sender for that recipient
depends on the phase of the moon.
.SH "SEE ALSO"
fastforward(1),
setforward(1),
newinclude(1),
printforward(1),
dot-qmail(5)


newaliases.c
#include "substdio.h"
#include "strerr.h"
#include "stralloc.h"
#include "getln.h"
#include "open.h"
#include "readwrite.h"
#include "token822.h"
#include "control.h"
#include "auto_qmail.h"
#include "case.h"
#include "cdbmss.h"

#define FATAL "newaliases: fatal: "

void nomem()
{
  strerr_die2x(111,FATAL,"out of memory");
}
void nulbyte()
{
  strerr_die2x(100,FATAL,"NUL bytes are not permitted");
}
void longaddress()
{
  strerr_die2x(100,FATAL,"addresses over 800 bytes are not permitted");
}
void writeerr()
{
  strerr_die2sys(111,FATAL,"unable to write to /etc/aliases.tmp: ");
}
void readerr()
{
  strerr_die2sys(111,FATAL,"unable to read /etc/aliases: ");
}
void die_control()
{
  strerr_die2sys(111,FATAL,"unable to read controls: ");
}

stralloc me = {0};
stralloc defaulthost = {0};
stralloc defaultdomain = {0};
stralloc plusdomain = {0};

void readcontrols()
{
  int r;
  int fddir;

  fddir = open_read(".");
  if (fddir == -1)
    strerr_die2sys(111,FATAL,"unable to open current directory: ");

  if (chdir(auto_qmail) == -1)
    strerr_die4sys(111,FATAL,"unable to chdir to ",auto_qmail,": ");

  r = control_readline(&me,"control/me");
  if (r == -1) die_control();
  if (!r) if (!stralloc_copys(&me,"me")) nomem();

  r = control_readline(&defaultdomain,"control/defaultdomain");
  if (r == -1) die_control();
  if (!r) if (!stralloc_copy(&defaultdomain,&me)) nomem();

  r = control_readline(&defaulthost,"control/defaulthost");
  if (r == -1) die_control();
  if (!r) if (!stralloc_copy(&defaulthost,&me)) nomem();

  r = control_readline(&plusdomain,"control/plusdomain");
  if (r == -1) die_control();
  if (!r) if (!stralloc_copy(&plusdomain,&me)) nomem();

  if (fchdir(fddir) == -1)
    strerr_die2sys(111,FATAL,"unable to set current directory: ");
}

stralloc target = {0};
stralloc fulltarget = {0};
stralloc instr = {0};

stralloc cbuf = {0};
token822_alloc toks = {0};
token822_alloc tokaddr = {0};
stralloc address = {0};

void gotincl()
{
  token822_reverse(&tokaddr);
  if (token822_unquote(&address,&tokaddr) != 1) nomem();
  tokaddr.len = 0;

  if (!address.len)
    strerr_die2x(111,FATAL,"empty :include: filenames not permitted");
  if (byte_chr(address.s,address.len,'\0') < address.len)
    strerr_die2x(111,FATAL,"NUL not permitted in :include: filenames");

  if ((address.s[0] != '.') && (address.s[0] != '/'))
    if (!stralloc_cats(&instr,"./")) nomem();
  if (!stralloc_cat(&instr,&address)) nomem();
  if (!stralloc_cats(&instr,".bin")) nomem();
  if (!stralloc_0(&instr)) nomem();
}

void gotaddr()
{
  int i;
  int j;
  int flaghasat;
 
  token822_reverse(&tokaddr);
  if (token822_unquote(&address,&tokaddr) != 1) nomem();

  if (!address.len)
    strerr_die2x(111,FATAL,"empty recipient addresses not permitted");
 
  flaghasat = 0;
  for (i = 0;i < tokaddr.len;++i)
    if (tokaddr.t[i].type == TOKEN822_AT)
      flaghasat = 1;
 
  tokaddr.len = 0;

  if (!address.len) return;

  if (!flaghasat)
    if (address.s[0] == '/') {
      if (!stralloc_0(&address)) nomem();
      strerr_die4x(111,FATAL,"file delivery ",address.s," not supported");
    }
  if (!flaghasat)
    if (address.s[0] == '|') {
      if (byte_chr(address.s,address.len,'\0') < address.len)
        strerr_die2x(111,FATAL,"NUL not permitted in program names");
      if (!stralloc_cats(&instr,"!")) nomem();
      if (!stralloc_catb(&instr,address.s + 1,address.len - 1)) nomem();
      if (!stralloc_0(&instr)) nomem();
      return;
    }


  if (target.len) {
    if (!stralloc_cats(&instr,"&")) nomem();
    if (!stralloc_cat(&instr,&fulltarget)) nomem();
    if (!stralloc_0(&instr)) nomem();
  }

  if (!flaghasat)
    if (!stralloc_cats(&address,"@")) nomem();

  if (!stralloc_copy(&target,&address)) nomem();
  if (!stralloc_copy(&fulltarget,&address)) nomem();

  if (fulltarget.s[fulltarget.len - 1] == '@')
     if (!stralloc_cat(&fulltarget,&defaulthost)) nomem();
  if (fulltarget.s[fulltarget.len - 1] == '+') {
    fulltarget.s[fulltarget.len - 1] = '.';
    if (!stralloc_cat(&fulltarget,&plusdomain)) nomem();
  }

  j = 0;
  for (i = 0;i < fulltarget.len;++i) if (fulltarget.s[i] == '@') j = i;
  for (i = j;i < fulltarget.len;++i) if (fulltarget.s[i] == '.') break;
  if (i == fulltarget.len) {
    if (!stralloc_cats(&fulltarget,".")) nomem();
    if (!stralloc_cat(&fulltarget,&defaultdomain)) nomem();
  }

  if (fulltarget.len > 800) longaddress();
  if (byte_chr(fulltarget.s,fulltarget.len,'\0') < fulltarget.len)
    strerr_die2x(111,FATAL,"NUL not permitted in recipient addresses");
}

stralloc line = {0};
stralloc newline = {0};
int match;

void parseerr()
{
  if (!stralloc_0(&line)) nomem();
  strerr_die3x(111,FATAL,"unable to parse this line: ",line.s);
}

void parseline()
{
  int wordok;
  struct token822 *t;
  struct token822 *beginning;
 
  switch(token822_parse(&toks,&line,&cbuf)) {
    case -1: nomem();
    case 0: parseerr();
  }

  beginning = toks.t;
  t = toks.t + toks.len;
  wordok = 1;
 
  if (!token822_readyplus(&tokaddr,1)) nomem();
  tokaddr.len = 0;
 
  while (t > beginning)
    switch((--t)->type) {
      case TOKEN822_SEMI:
        break; /*XXX*/
      case TOKEN822_COLON:
        if (t >= beginning + 2)
          if (t[-2].type == TOKEN822_COLON)
            if (t[-1].type == TOKEN822_ATOM)
              if (t[-1].slen == 7)
                if (!byte_diff(t[-1].s,7,"include")) {
                  gotincl();
                  t -= 2;
                }
        break; /*XXX*/
      case TOKEN822_RIGHT:
        if (tokaddr.len) gotaddr();
        while ((t > beginning) && (t[-1].type != TOKEN822_LEFT))
          if (!token822_append(&tokaddr,--t)) nomem();
        gotaddr();
        if (t <= beginning) parseerr();
        --t;
        while ((t > beginning) && ((t[-1].type == TOKEN822_COMMENT) || (t[-1].type == TOKEN822_ATOM) || (t[-1].type == TOKEN822_QUOTE) || (t[-1].type == TOKEN822_AT) || (t[-1].type == TOKEN822_DOT)))
          --t;
        wordok = 0;
        continue;
      case TOKEN822_ATOM: case TOKEN822_QUOTE: case TOKEN822_LITERAL:
        if (!wordok) if (tokaddr.len) gotaddr();
        wordok = 0;
        if (!token822_append(&tokaddr,t)) nomem();
        continue;
      case TOKEN822_COMMENT:
        /* comment is lexically a space; shouldn't affect wordok */
        break;
      case TOKEN822_COMMA:
        if (tokaddr.len) gotaddr();
        wordok = 1;
        break;
      default:
        wordok = 1;
        if (!token822_append(&tokaddr,t)) nomem();
        continue;
    }
  if (tokaddr.len) gotaddr();
}

char inbuf[1024];
substdio ssin;
struct cdbmss cdbmss;
stralloc key = {0};

void doit()
{
  if (!instr.len) {
    if (target.len) parseerr();
    return;
  }

  if (!target.len) parseerr();

  if (stralloc_starts(&target,"owner-")) {
    if (!stralloc_copys(&key,"?")) nomem();
    if (!stralloc_catb(&key,target.s + 6,target.len - 6)) nomem();
    case_lowerb(key.s,key.len);
    if (cdbmss_add(&cdbmss,key.s,key.len,fulltarget.s,fulltarget.len) == -1) writeerr();
  }

  if (!stralloc_copys(&key,":")) nomem();
  if (!stralloc_cat(&key,&target)) nomem();
  case_lowerb(key.s,key.len);
  if (cdbmss_add(&cdbmss,key.s,key.len,instr.s,instr.len) == -1) writeerr();
}

int main()
{
  int fd;

  umask(033);
  readcontrols();

  fd = open_read("/etc/aliases");
  if (fd == -1) readerr();
  substdio_fdbuf(&ssin,read,fd,inbuf,sizeof(inbuf));

  fd = open_trunc("/etc/aliases.tmp");
  if (fd == -1) strerr_die2sys(111,FATAL,"unable to create /etc/aliases.tmp: ");
  if (cdbmss_start(&cdbmss,fd) == -1) writeerr();

  if (!stralloc_copys(&line,"")) nomem();

  for (;;) {
    if (getln(&ssin,&newline,&match,'\n') != 0) readerr();

    if (match && (newline.s[0] == '\n')) continue;

    if (match && ((newline.s[0] == ' ') || (newline.s[0] == '\t'))) {
      if (!stralloc_cat(&line,&newline)) nomem();
      continue;
    }

    if (line.len)
      if (line.s[0] != '#') {
        if (!stralloc_copys(&target,"")) nomem();
        if (!stralloc_copys(&fulltarget,"")) nomem();
        if (!stralloc_copys(&instr,"")) nomem();
        parseline();
	doit();
      }

    if (!match) break;
    if (!stralloc_copy(&line,&newline)) nomem();
  }

  if (cdbmss_finish(&cdbmss) == -1) writeerr();
  if (fsync(fd) == -1) writeerr();
  if (close(fd) == -1) writeerr(); /* NFS stupidity */

  if (rename("/etc/aliases.tmp","/etc/aliases.cdb") == -1)
    strerr_die2sys(111,FATAL,"unable to move /etc/aliases.tmp to /etc/aliases.cdb: ");
  
  _exit(0);
}
